n3fit/src/n3fit/hyper_optimization/hyper_scan.py

```diff
@@ -114,28 +115,64 @@ def hyper_scan_wrapper(replica_path_set, model_trainer, hyperscanner, max_evals=
     """
     # Tell the trainer we are doing hpyeropt
     model_trainer.set_hyperopt(True, keys=hyperscanner.hyper_keys, status_ok=hyperopt.STATUS_OK)
+
     # Generate the trials object
-    trials = FileTrials(replica_path_set, parameters=hyperscanner.as_dict())
+    if hyperscanner.parallel_hyperopt:
+        # Instantiate `MongoFileTrials`
+        # Mongo database should have already been initiated at this point
+        trials = MongoFileTrials(
+            replica_path_set,
+            db_host=hyperscanner.db_host,
+            db_port=hyperscanner.db_port,
+            db_name=hyperscanner.db_name,
+            num_workers=hyperscanner.num_mongo_workers,
+            parameters=hyperscanner.as_dict(),
+        )
+    else:
+        # Instantiate `FileTrials`
+        trials = FileTrials(replica_path_set, parameters=hyperscanner.as_dict())
+
     # Initialize seed for hyperopt
     trials.rstate = np.random.default_rng(HYPEROPT_SEED)
 
-    # For restarts, reset the state of `FileTrials` saved in the pickle file
-    if hyperscanner.restart_hyperopt:
+    # For sequential hyperopt restarts, reset the state of `FileTrials` saved in the pickle file
+    if not hyperscanner.parallel_hyperopt and hyperscanner.restart_hyperopt:
         pickle_file_to_load = f"{replica_path_set}/tries.pkl"
         log.info("Restarting hyperopt run using the pickle file %s", pickle_file_to_load)
         trials = FileTrials.from_pkl(pickle_file_to_load)
 
-    # Perform the scan
-    best = hyperopt.fmin(
-        fn=model_trainer.hyperparametrizable,
-        space=hyperscanner.as_dict(),
-        algo=hyperopt.tpe.suggest,
-        max_evals=max_evals,
-        show_progressbar=False,
-        trials=trials,
-        rstate=trials.rstate,
-        trials_save_file=trials.pkl_file,
-    )
+    # Call to hyperopt.fmin
+    if hyperscanner.parallel_hyperopt:
+        # Launch mongo workers
+        trials.start_mongo_workers()
+
+        # Perform the scan in parallel
+        best = hyperopt.fmin(
+            fn=model_trainer.hyperparametrizable,
+            space=hyperscanner.as_dict(),
+            algo=hyperopt.tpe.suggest,
+            max_evals=max_evals,
+            show_progressbar=True,
+            trials=trials,
+            rstate=trials.rstate,
+            max_queue_len=trials.num_workers,
+        )
+
+        # Stop mongo workers
+        trials.stop_mongo_workers()
+    else:
+        # Perform the scan sequentially
+        best = hyperopt.fmin(
+            fn=model_trainer.hyperparametrizable,
+            space=hyperscanner.as_dict(),
+            algo=hyperopt.tpe.suggest,
+            max_evals=max_evals,
+            show_progressbar=False,
+            trials=trials,
+            rstate=trials.rstate,
+            trials_save_file=trials.pkl_file,
+        )
+
```

Comment:
```suggestion
       # Call to hyperopt.fmin
    fmin_args = dict(
        fn=model_trainer.hyperparametrizable,
        space=hyperscanner.as_dict(),
        algo=hyperopt.tpe.suggest,
        max_evals=max_evals,
        trials=trials,
        rstate=trials.rstate,
    )
    if hyperscanner.parallel_hyperopt:
        trials.start_mongo_workers()
        best = hyperopt.fmin(**fmin_args, show_progressbar=True, max_queue_len=trials.num_workers)
        trials.stop_mongo_workers()
    else:
        best = hyperopt.fmin(**fmin_args, show_progressbar=False, trials_save_file=trials.pkl_file)
```

Perhaps something like this instead, to make it clear what the two branches have in common?

Reply:
Looks nice. Thanks for your suggestion. I will accept it.